File: src/core/truth-engine/core/ontology.ts
 /**
  * TRUTH ENGINE - CORE ONTOLOGY
  * 
  * The 9 permitted base classes.
  * NO OTHER TYPES MAY EXIST.
  */
 
 /**
  * BASE CLASS ENUMERATION (LOCKED)
  */
 export const BASE_CLASSES = [
   'Entity',
   'Attribute', 
   'Relation',
   'Event',
   'Measure',
   'Source',
   'Schema',
   'Version',
   'Conclusion',
 ] as const;
 
 export type BaseClass = typeof BASE_CLASSES[number];
 
 /**
  * TEMPORAL ENVELOPE (REQUIRED FOR ALL OBJECTS)
  */
 export interface TemporalEnvelope {
   readonly observed_at: string;   // ISO 8601
   readonly valid_from: string;    // ISO 8601
   readonly valid_to: string | null; // ISO 8601 or null for ongoing
 }
 
 /**
  * SOURCE ENVELOPE (REQUIRED FOR ALL DATA)
  */
 export interface SourceEnvelope {
   readonly source_id: string;
   readonly source_version: string;
   readonly source_accessed_at: string;
 }
 
 /**
  * UNCERTAINTY ENVELOPE (REQUIRED FOR ALL MEASURES)
  */
 export interface UncertaintyEnvelope {
   readonly confidence_interval: [number, number];
   readonly coverage: number; // 0-1
   readonly methodology_note: string;
 }
 
 /**
  * CORE OBJECT - Base for all truth engine objects
  */
 export interface CoreObject {
   readonly id: string;
   readonly baseClass: BaseClass;
   readonly schema_id: string;
   readonly created_at: string;
   readonly created_by: string;
 }
 
 /**
  * ENTITY - A thing that exists
  */
 export interface Entity extends CoreObject {
   readonly baseClass: 'Entity';
   readonly entity_type: string;
   readonly name: string;
   readonly identifiers: Record<string, string>;
 }
 
 /**
  * SOURCE - Origin of data
  */
 export interface Source extends CoreObject {
   readonly baseClass: 'Source';
   readonly name: string;
   readonly organization: string;
   readonly url: string | null;
   readonly reliability_score: number; // 0-1
   readonly methodology_url: string | null;
 }
 
 /**
  * SCHEMA - Definition of a measure
  */
 export interface Schema extends CoreObject {
   readonly baseClass: 'Schema';
   readonly name: string;
   readonly definition: string;
   readonly unit: string;
   readonly dimension: string;
   readonly version: number;
   readonly supersedes: string | null;
 }
 
 /**
  * MEASURE - A quantified observation
  */
 export interface Measure extends CoreObject {
   readonly baseClass: 'Measure';
   readonly schema_id: string;
   readonly entity_id: string;
   readonly value: number;
   readonly unit: string;
   readonly temporal: TemporalEnvelope;
   readonly source: SourceEnvelope;
   readonly uncertainty: UncertaintyEnvelope;
 }
 
 /**
  * Type guards
  */
 export function isValidBaseClass(value: unknown): value is BaseClass {
   return typeof value === 'string' && BASE_CLASSES.includes(value as BaseClass);
 }
 
 export function isMeasure(obj: CoreObject): obj is Measure {
   return obj.baseClass === 'Measure';
 }
 
 export function isSource(obj: CoreObject): obj is Source {
   return obj.baseClass === 'Source';
 }
 
 export function isSchema(obj: CoreObject): obj is Schema {
   return obj.baseClass === 'Schema';
 }
 
 export function isEntity(obj: CoreObject): obj is Entity {
   return obj.baseClass === 'Entity';
 }

File: src/core/truth-engine/python-reference/ontology.py
 """
 TRUTH ENGINE - CORE ONTOLOGY
 
 The 9 permitted base classes.
 NO OTHER TYPES MAY EXIST.
 
 This is the Python reference implementation for standalone deployment.
 The TypeScript version in ../core/ontology.ts is functionally identical.
 """
 
 from enum import Enum
 from dataclasses import dataclass
 from typing import Optional, Dict, Tuple
 from datetime import datetime
 
 
 class BaseType(Enum):
     """The 9 permitted base classes. No other types may exist."""
     ENTITY = "entity"
     ATTRIBUTE = "attribute"
     RELATION = "relation"
     EVENT = "event"
     MEASURE = "measure"
     SOURCE = "source"
     SCHEMA = "schema"
     VERSION = "version"
     CONCLUSION = "conclusion"
 
 
 @dataclass(frozen=True)
 class TemporalEnvelope:
     """Required for all objects."""
     observed_at: str  # ISO 8601
     valid_from: str   # ISO 8601
     valid_to: Optional[str] = None  # ISO 8601 or None for ongoing
 
 
 @dataclass(frozen=True)
 class SourceEnvelope:
     """Required for all data."""
     source_id: str
     source_version: str
     source_accessed_at: str
 
 
 @dataclass(frozen=True)
 class UncertaintyEnvelope:
     """Required for all measures."""
     confidence_interval: Tuple[float, float]
     coverage: float  # 0-1
     methodology_note: str
 
 
 @dataclass(frozen=True)
 class CoreObject:
     """Base for all truth engine objects."""
     id: str
     base_class: BaseType
     schema_id: str
     created_at: str
     created_by: str
 
 
 @dataclass(frozen=True)
 class Entity(CoreObject):
     """A thing that exists."""
     entity_type: str
     name: str
     identifiers: Dict[str, str]
 
 
 @dataclass(frozen=True)
 class Source(CoreObject):
     """Origin of data."""
     name: str
     organization: str
     url: Optional[str]
     reliability_score: float  # 0-1
     methodology_url: Optional[str]
 
 
 @dataclass(frozen=True)
 class Schema(CoreObject):
     """Definition of a measure."""
     name: str
     definition: str
     unit: str
     dimension: str
     version: int
     supersedes: Optional[str]
 
 
 @dataclass(frozen=True)
 class Measure(CoreObject):
     """A quantified observation."""
     entity_id: str
     value: float
     unit: str
     temporal: TemporalEnvelope
     source: SourceEnvelope
     uncertainty: UncertaintyEnvelope
 
 
 def is_valid_base_class(value: str) -> bool:
     """Type guard for base class validation."""
     try:
         BaseType(value)
         return True
     except ValueError:
         return False